server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -372,24 +376,22 @@ private synchronized DocumentMapper internalMerge(DocumentMapper mapper, MergeRe
 
         ContextMapping.validateContextPaths(indexSettings.getIndexVersionCreated(), fieldMappers, newFieldTypes::get);
 
-        if (reason == MergeReason.MAPPING_UPDATE || reason == MergeReason.MAPPING_UPDATE_PREFLIGHT) {
-            // this check will only be performed on the master node when there is
-            // a call to the update mapping API. For all other cases like
-            // the master node restoring mappings from disk or data nodes
-            // deserializing cluster state that was sent by the master node,
-            // this check will be skipped.
+        if (reason != MergeReason.MAPPING_RECOVERY) {
+            // this check will only be performed on the master node when an index is created, or
+            // there is a call to the update mapping API. For all other cases like the master node
+            // restoring mappings from disk or data nodes deserializing cluster state that was sent
+            // by the master node, this check will be skipped.
             // Also, don't take metadata mappers into account for the field limit check
             checkTotalFieldsLimit(objectMappers.size() + fieldMappers.size() - metadataMappers.length
                 + fieldAliasMappers.size() );
             checkFieldNameSoftLimit(objectMappers, fieldMappers, fieldAliasMappers);
         }
 
-        if (reason == MergeReason.MAPPING_UPDATE || reason == MergeReason.MAPPING_UPDATE_PREFLIGHT) {
-            // this check will only be performed on the master node when there is
-            // a call to the update mapping API. For all other cases like
-            // the master node restoring mappings from disk or data nodes
-            // deserializing cluster state that was sent by the master node,
-            // this check will be skipped.
+        if (reason != MergeReason.MAPPING_RECOVERY) {
```

Comment:
I think this block can be folded into the same check a few lines above rather than having two blocks using the same `if` statement

Reply:
👍